dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ReturnTypesResolver.java

```diff
@@ -82,15 +83,29 @@ private Optional<TypeName> orDefaultDeserializerFactory(
                 context.reportError("When returning raw Response, remember to add @MustBeClosed annotation", element);
                 return Optional.empty();
             }
-            return Optional.of(context.getTypeName(ResponseDeserializer.class));
+            return Optional.of(context.getTypeMirror(ResponseDeserializer.class));
+        } else if (isInputStreamType(returnType)) {
+            if (!hasMustBeClosed) {
+                context.reportError("When returning InputStream, remember to add @MustBeClosed annotation", element);
+                return Optional.empty();
+            }
+            return Optional.of(context.getTypeMirror(InputStreamDeserializer.class));
         }
-        return Optional.of(context.getTypeName(Json.class));
+        return Optional.of(context.getTypeMirror(Json.class));
     }
 
     private boolean isResponseType(TypeMirror type) {
         return context.isSameTypes(type, Response.class);
     }
 
+    private boolean isInputStreamType(TypeMirror type) {
+        return context.isSameTypes(type, InputStream.class);
+    }
+
+    private boolean usesBodySerde(TypeMirror type) {
+        return context.isSameTypes(type, InputStreamDeserializer.class);
```

Comment:
I tried to have something a bit more generic by making a `BodySerDeBasedDeserializer<T>` class that `InputStreamDeserializer` extends and use `context.isAssignable` here, but that didn't seem to work.

I haven't figured out a nicer way to make it possible to rely on the BodySerDe's existing InputStreamDeserializer either